FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+                 * label and type. The intent of this loop is to try to delete all of
+                 * them. However, to avoid getting stuck, we won't try to find another
+                 * object of the same label/type if the previous delete failed. */
+                if( xResult == CKR_OK )
+                {
+                    xResult = xFindObjectWithLabelAndClass( xSession, ( char * ) pxLabelPtr,
+                                                            strlen( ( char * ) pxLabelPtr ),
+                                                            pxClass[ xIndex ], &xObjectHandle );
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionPrivateECKey( CK_SESSION_HANDLE xSession,
+                                       const char * pcLabel,
+                                       mbedtls_pk_context * pxMbedPkContext )
+{
+    CK_RV xResult = CKR_OK;
+    CK_FUNCTION_LIST_PTR xFunctionList = NULL;
+    CK_BYTE * pxDPtr = NULL;        /* Private value D. */
+    CK_BYTE * pxEcParamsPtr = NULL; /* DER-encoding of an ANSI X9.62 Parameters value */
+    int xMbedResult = 0;
+    CK_BBOOL xTrueObject = CK_TRUE;
+    CK_KEY_TYPE xPrivateKeyType = CKK_EC;
+    CK_OBJECT_CLASS xPrivateKeyClass = CKO_PRIVATE_KEY;
+    CK_OBJECT_HANDLE xObjectHandle = CK_INVALID_HANDLE;
+    mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedPkContext->pk_ctx;
+
+    xResult = C_GetFunctionList( &xFunctionList );
+
+    if( xResult != CKR_OK )
+    {
+        LogError( ( "Could not get a PKCS #11 function pointer." ) );
+    }
+    else
+    {
+        pxDPtr = ( CK_BYTE * ) malloc( EC_D_LENGTH );
+
+        if( pxDPtr == NULL )
+        {
+            xResult = CKR_HOST_MEMORY;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        xMbedResult = mbedtls_mpi_write_binary( &( pxKeyPair->d ), pxDPtr, EC_D_LENGTH );
+
+        if( xMbedResult != 0 )
+        {
+            LogError( ( "Failed to parse EC private key components." ) );
+            xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        if( pxKeyPair->grp.id == MBEDTLS_ECP_DP_SECP256R1 )
+        {
+            pxEcParamsPtr = ( CK_BYTE * ) ( "\x06\x08" MBEDTLS_OID_EC_GRP_SECP256R1 );
+        }
+        else
+        {
+            xResult = CKR_CURVE_NOT_SUPPORTED;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        CK_ATTRIBUTE privateKeyTemplate[] =
+        {
+            { CKA_CLASS,     NULL /* &privateKeyClass*/, sizeof( CK_OBJECT_CLASS )      },
+            { CKA_KEY_TYPE,  NULL /* &privateKeyType*/,  sizeof( CK_KEY_TYPE )          },
+            { CKA_LABEL,     ( void * ) pcLabel,         ( CK_ULONG ) strlen( pcLabel ) },
```

Comment:
```suggestion
            { CKA_LABEL,     ( void * ) pcLabel,         ( CK_ULONG ) strnlen( pcLabel, MAX_LEN? ) },
```

Reply:
Resolved in this commit (along with other labels with similar issues): https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/286f3f5c78e9b0ed419b68e7325f3f02683bc5fc